### PR TITLE
Bug 1078209 - Get next N fixes

### PR DESCRIPTION
**Pull request by camd**
This addresses [Bug 1078209](https://bugzilla.mozilla.org/show_bug.cgi?id=1078209)

Here are the workflows this change facilitates:

1. If you have the ``revision`` value set (so you're viewing a single revision) hitting "Next N" will change the param from ``revision`` to ``tochange`` and set the ``fromchange`` to the latest loaded (earliest) resultset.

2. after loading default of latest 10, click next 10 will load them and set only the ``fromchange`` value to the latest revision loaded.  Hit next N again and it will update the ``fromchange`` to the latest again.  
Note: ``tochange`` is not updated in this scenario, so that new resultsets will continue to be loaded by default

3. The "action" menu (at the right end of each resultset) has two new menu items:

    1. "Set as tochange" will set the ``tochange`` param to that revision and will therefore stop loading any new resultsets.

    2. "Set as fromchange`` will set the ``fromchange`` param to that revision.  Hitting "Next N" will update that value as you go.

Found, but not addressed in this PR:
[Bug 1168896](https://bugzilla.mozilla.org/show_bug.cgi?id=1168896)

[Review on Reviewable](https://reviewable.io/reviews/mozilla/treeherder/560)

### CODE REVIEW

**Comment by tojonmz** [comment]
For what its worth I tested this branch locally two days ago and it seemed to behave nicely, in accordance with Ed's [requested behaviors] (https://bugzilla.mozilla.org/show_bug.cgi?id=1078209#c0) in the bug.

The only thing I noticed was that when doing a get-next from a single resultset view the 'tochange' precedes the 'fromchange' in the url. It might feel a bit backwards reading the url, but on the plus side when you switch to the range view, that original SHA doesn't jump around in the url. It stays relatively in the same place.

File: ui/js/controllers/jobs.js [hunk 1]
```diff
@@ -18,8 +18,26 @@ treeherderApp.controller('JobsCtrl', [
...
+            }
+            ThResultSetStore.fetchResultSets($scope.repoName, count, keepFilters).
+                then(function() {
+
+                    // since we fetched more resultsets, we need to persist the
+                    // resultset state in the URL.
+                    $rootScope.skipNextPageReload = true;
+                    console.log("setting location");
+                    var rsArray = ThResultSetStore.getResultSetsArray($scope.repoName)
+//                    $location.search('tochange', _.first(rsArray).revision);
```
Comment (tojonmz) [inline ui/js/controllers/jobs.js]: Noticed this earlier, for removal(?).

File: ui/js/controllers/main.js [hunk 2]
```diff
@@ -332,13 +332,18 @@ treeherderApp.controller('MainCtrl', [
             var defaulting = newReloadTriggerParams.repo === thDefaultRepo &&
                              !$scope.cachedReloadTriggerParams.repo;
 
+            console.log("skipPageReload before check", $scope.skipNextPageReload);
             if (!defaulting && $scope.cachedReloadTriggerParams &&
-                !_.isEqual(newReloadTriggerParams, $scope.cachedReloadTriggerParams)) {
+                !_.isEqual(newReloadTriggerParams, $scope.cachedReloadTriggerParams) &&
+                !$rootScope.skipNextPageReload) {
 
+//                console.log("oh man, I was going to reload the page", $scope.skipPageReload);
```
Comment (tojonmz) [inline ui/js/controllers/main.js]: And here also.

**Comment by camd** [comment]
I don't really have much control over the order of the params, tbh.  Though, since the resultsets are in descending order, maybe that could still make sense, in a manner of speaking.  Like "Darth Vader killed your father, Luke..." makes sense, from a certain perspective.

**Comment by edmorley** [comment]
Really liking the workflow with these changes - adding the fromchange/tochange to the push menu was a great idea :-)

The only problem I found is that after opening inbound (with no params) and then choosing "set as tochange" on the second push from the top, resulted in ~100 pushes being opened - rather than the 10 or so I would have expected?

---

Review status: all files reviewed, 2 unresolved discussions, all commit checks successful.
Reviewed files:
- ui/js/controllers/jobs.js @ r3
- ui/js/controllers/main.js @ r3
- ui/js/models/resultset.js @ r3
- ui/js/models/resultsets_store.js @ r3
- ui/partials/main/jobs.html @ r3
- ui/partials/main/thActionButton.html @ r2

---

<sup>**[ui/js/controllers/jobs.js, line 22 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/560#-JpxwI11a5g7qhe75pky-r1-37)** ([raw file](https://github.com/mozilla/treeherder/blob/0273e8cbb09167235d299717260a2edf0b8063b8/ui/js/controllers/jobs.js#L22)):</sup>
jfrench, I don't follow?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/560)

**Comment by camd** [comment]
Review status: all files reviewed, 2 unresolved discussions, all commit checks successful.

---

<sup>**[ui/js/controllers/jobs.js, line 22 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/560#-JpxwI11a5g7qhe75pky-r1-37)** ([raw file](https://github.com/mozilla/treeherder/blob/0273e8cbb09167235d299717260a2edf0b8063b8/ui/js/controllers/jobs.js#L22)):</sup>
it was a comment I'd left in an earlier version.  Removed now.  I guess reviewable didn't notice the line had changed.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/560)

**Comment by camd** [comment]
Review status: all files reviewed, 2 unresolved discussions, all commit checks successful.

---

<sup>**[ui/js/controllers/main.js, line 348 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/560#-JpxwI14O6hs14IlP-OF-r1-340)** ([raw file](https://github.com/mozilla/treeherder/blob/0273e8cbb09167235d299717260a2edf0b8063b8/ui/js/controllers/main.js#L348)):</sup>
removed

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/560)

**Comment by tojonmz** [comment]
Review status: all files reviewed, 2 unresolved discussions, all commit checks successful.

---

<sup>**[ui/js/controllers/jobs.js, line 22 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/560#-JpxwI11a5g7qhe75pky-r1-37)** ([raw file](https://github.com/mozilla/treeherder/blob/0273e8cbb09167235d299717260a2edf0b8063b8/ui/js/controllers/jobs.js#L22)):</sup>
Yup, that must have been the case.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/560)

**Comment by camd** [comment]
Yeah, that's a good point.  I've gone back and forth a bit with the expected behavior on those settings.  I think with the case of ``fromchange`` only, 100 max is the right call.  But with ``tochange`` only, then it should just get 10, like you say.  

---

Review status: 5 of 6 files reviewed, 2 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/560)

**Comment by tojonmz** [comment]
I tested this latest branch locally and it seems great in all workflows I could think of. The only items were as discussed in channel; potentially non-techy names for the menus later (if we can think of any better ones) and the possible need to clear a "to" of a from/to, which if it wasn't the latest push I guess requires manual removal in the url bar.

Looks great camd!

---

Review status: 4 of 6 files reviewed, 2 unresolved discussions, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/560)

**Comment by edmorley** [comment]
Review status: all files reviewed, 2 unresolved discussions, all commit checks successful.
Reviewed files:
- ui/js/models/resultset.js @ r4
- ui/partials/main/thActionButton.html @ r4

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/560)

**Pull request merged**